web/src/features/datasets/server/service.ts

```diff
@@ -0,0 +1,329 @@
+          : observation?.cost
+            ? new Decimal(observation.cost)
+            : new Decimal(0),
+        countRunItems: run_items.length,
+        name: run.run_name,
+        description: run.run_description,
+        metadata: run.run_metadata,
+        createdAt: run.run_created_at,
+        updatedAt: run.run_updated_at,
+        avgLatency: trace?.latency ?? observation?.latency ?? 0,
+        scores: aggregateScores(scores.filter((s) => s.run_id === run.run_id)),
+      };
+    });
+
+    return enrichedRuns;
+  } catch (e) {
+    logger.error("Failed to fetch dataset runs from clickhouse", e);
+    await deleteTempTableInClickhouse(tableName);
+    throw e;
+  }
+};
+
+export const insertPostgresDatasetRunsIntoClickhouse = async (
+  runs: PostgresDatasetRun[],
+  tableName: string,
+  projectId: string,
+  datasetId: string,
+  clickhouseSession: string,
+) => {
+  const rows = runs.flatMap((run) =>
+    run.run_items.map((item) => ({
+      project_id: projectId,
+      run_id: run.run_id,
+      run_item_id: item.ri_id,
+      dataset_id: datasetId,
+      trace_id: item.trace_id,
+      observation_id: item.observation_id,
+    })),
+  );
+
+  await clickhouseClient({ session_id: clickhouseSession }).insert({
+    table: tableName,
+    values: rows,
+    format: "JSONEachRow",
+  });
+};
+
+export const createTempTableInClickhouse = async (
+  tableName: string,
+  clickhouseSession: string,
+) => {
+  const query = `
+      CREATE TEMPORARY TABLE IF NOT EXISTS ${tableName}
+      (
+          project_id String,    
+          run_id String,  
+          run_item_id String,
+          dataset_id String,      
+          trace_id String,
+          observation_id Nullable(String)
+      )  ENGINE = Memory
+
+  `;
+  await commandClickhouse({
+    query,
+    params: { tableName },
+    clickhouseConfigs: { session_id: clickhouseSession },
+  });
+};
+
+export const deleteTempTableInClickhouse = async (tableName: string) => {
+  const query = `
+      DROP TABLE IF EXISTS ${tableName}
+  `;
+  await commandClickhouse({ query, params: { tableName } });
+};
+
+export const getDatasetRunsFromPostgres = async (
+  input: DatasetRunsTableInput,
+) => {
+  return await prisma.$queryRaw<PostgresDatasetRun[]>(
+    Prisma.sql`
+      SELECT
+        runs.id as run_id,
+        runs.name as run_name,
+        runs.description as run_description,
+        runs.metadata as run_metadata,
+        runs.created_at as run_created_at,
+        runs.updated_at as run_updated_at,
+        JSON_AGG(JSON_BUILD_OBJECT(
+          'trace_id', ri.trace_id,
+          'observation_id', ri.observation_id,
+          'ri_id', ri.id
+        )) AS run_items
+      FROM
+        datasets d
+        JOIN dataset_runs runs ON d.id = runs.dataset_id AND d.project_id = runs.project_id
+        JOIN dataset_run_items ri ON ri.dataset_run_id = runs.id
+          AND ri.project_id = runs.project_id
+      WHERE
+        d.id = ${input.datasetId}
```

Comment:
This prisma syntax still freaks me out a bit. There is no need to wrap those inputs again in any way, right?

Reply:
No, not in this case. This will be sent as a SQL object as prepared statement. If I would have used $queryRawUnsafe, this would be an issue.